packages/evm/contracts/budgets/TransparentBudget.sol

```diff
@@ -0,0 +1,339 @@
+    function getFungibleAmountAndKey(address asset)
+        internal
+        pure
+        returns (LibTransient.TUint256 storage p, bytes32 key)
+    {
+        key = bytes32(uint256(uint160(asset)));
+        p = LibTransient.tUint256(key);
+    }
+
+    function getERC1155AmountAndKey(address asset, uint256 tokenId)
+        internal
+        pure
+        returns (LibTransient.TUint256 storage p, bytes32 key)
+    {
+        key = keccak256(abi.encodePacked(asset, tokenId));
+        p = LibTransient.tUint256(key);
+    }
+
+    /// @inheritdoc ABudget
+    /// @notice Reclaims assets from the budget
+    /// @param data_ The packed {Transfer} request
+    /// @return True if the request was successful
+    /// @dev Only the owner can directly reclaim assets from the budget
+    /// @dev If the amount is zero, the entire balance of the asset will be transferred to the receiver
+    /// @dev If the asset transfer fails, the reclamation will revert
+    function clawback(bytes calldata data_) external virtual override onlyOwner returns (uint256) {
+        uint256 amount = 0;
+        Transfer memory request = abi.decode(data_, (Transfer));
+        if (request.assetType == AssetType.ETH || request.assetType == AssetType.ERC20) {
+            FungiblePayload memory payload = abi.decode(request.data, (FungiblePayload));
+            _transferFungible(
+                request.asset, request.target, payload.amount == 0 ? available(request.asset) : payload.amount
+            );
+            amount = payload.amount;
+        } else if (request.assetType == AssetType.ERC1155) {
+            ERC1155Payload memory payload = abi.decode(request.data, (ERC1155Payload));
+            _transferERC1155(
+                request.asset,
+                request.target,
+                payload.tokenId,
+                payload.amount == 0 ? IERC1155(request.asset).balanceOf(address(this), payload.tokenId) : payload.amount,
+                payload.data
+            );
+            amount = payload.amount;
+        }
+
+        return amount;
+    }
+
+    /// @inheritdoc ABudget
+    /// @notice Disburses assets from the budget to a single recipient
+    /// @param data_ The packed {Transfer} request
+    /// @return True if the disbursement was successful
+    /// @dev If the asset transfer fails, the disbursement will revert
+    function disburse(bytes calldata data_) public virtual override returns (bool) {
+        Transfer memory request = abi.decode(data_, (Transfer));
+        if (request.assetType == AssetType.ERC20 || request.assetType == AssetType.ETH) {
+            FungiblePayload memory payload = abi.decode(request.data, (FungiblePayload));
+
+            (LibTransient.TUint256 storage p,) = getFungibleAmountAndKey(request.asset);
+            uint256 avail = p.get();
+            if (payload.amount > avail) {
+                revert InsufficientFunds(request.asset, avail, payload.amount);
+            }
+
+            p.dec(payload.amount);
```

Comment:
And here's where we unlatch - presumption being this should zero it on after the core call to disburse then the check in the internal budgets `createBoost` call will fail if it doesn't unlatch/zero-out correctly.

Reply:
Yeah the beauty here is that since the accounting is entirely tracked by internal state (no `balanceOf` calls) if reentrancy is attempted on incentive disbursements will 100% not zero out.